File: paimeng-ai-code-agent/src/codegen/parsing.ts
// 代码块解析（Issue #8）：从 Python Agent 的 app/services/codegen/parsing.py 按语义移植，
// 正则与旧实现逐字对齐，输出语义等价。工作流把 LLM 流式产出解析为 HTML/CSS/JS 文件集后落盘。
const HTML_PATTERN = /```html\s*\n([\s\S]*?)```/i
const CSS_PATTERN = /```css\s*\n([\s\S]*?)```/i
const JS_PATTERN = /```(?:js|javascript)\s*\n([\s\S]*?)```/i

export interface HtmlCodeResult {
  htmlCode: string
  description: string
}

export interface MultiFileCodeResult {
  htmlCode: string
  cssCode: string
  jsCode: string
  description: string
}

// 解析单文件 HTML：提取 ```html 代码块内容；无代码块时把整个内容作为 HTML（对齐 Python parse_html_code）
export function parseHtmlCode(codeContent: string): HtmlCodeResult {
  const matched = HTML_PATTERN.exec(codeContent)?.[1]?.trim()
  if (matched) {
    return { htmlCode: matched, description: '' }
  }
  return { htmlCode: codeContent.trim(), description: '' }
}

// 解析多文件代码：分别提取 html/css/js 三块（缺失时对应字段为空，对齐 Python parse_multi_file_code）
export function parseMultiFileCode(codeContent: string): MultiFileCodeResult {
  const extract = (pattern: RegExp): string => pattern.exec(codeContent)?.[1]?.trim() ?? ''
  return {
    htmlCode: extract(HTML_PATTERN),
    cssCode: extract(CSS_PATTERN),
    jsCode: extract(JS_PATTERN),
    description: '',
  }
}

// 解析结果 → 「相对路径 → 内容」文件集（跳过空白内容；对齐 Python to_files）
export function toFiles(result: HtmlCodeResult | MultiFileCodeResult): Record<string, string> {
  const files: Record<string, string> = {}
  if ('cssCode' in result) {
    if (result.htmlCode) files['index.html'] = result.htmlCode
    if (result.cssCode) files['style.css'] = result.cssCode
    if (result.jsCode) files['script.js'] = result.jsCode
  } else if (result.htmlCode) {
    files['index.html'] = result.htmlCode
  }
  return files
}

File: paimeng-ai-code-rag/app/services/codegen/parsing.py
import re
from dataclasses import dataclass, field

HTML_PATTERN = re.compile(r"```html\s*\n([\s\S]*?)```", re.IGNORECASE)
CSS_PATTERN = re.compile(r"```css\s*\n([\s\S]*?)```", re.IGNORECASE)
JS_PATTERN = re.compile(r"```(?:js|javascript)\s*\n([\s\S]*?)```", re.IGNORECASE)


@dataclass
class HtmlCodeResult:
    html_code: str = ""
    description: str = ""


@dataclass
class MultiFileCodeResult:
    html_code: str = ""
    css_code: str = ""
    js_code: str = ""
    description: str = ""


def parse_html_code(code_content: str) -> HtmlCodeResult:

    matcher = HTML_PATTERN.search(code_content)
    if matcher and matcher.group(1).strip():
        return HtmlCodeResult(html_code=matcher.group(1).strip())
    return HtmlCodeResult(html_code=code_content.strip())


def parse_multi_file_code(code_content: str) -> MultiFileCodeResult:

    result = MultiFileCodeResult()

    def extract(pattern: re.Pattern[str]) -> str:
        matcher = pattern.search(code_content)
        return matcher.group(1).strip() if matcher else ""

    result.html_code = extract(HTML_PATTERN)
    result.css_code = extract(CSS_PATTERN)
    result.js_code = extract(JS_PATTERN)
    return result


def to_files(result: HtmlCodeResult | MultiFileCodeResult) -> dict[str, str]:

    files: dict[str, str] = {}
    if isinstance(result, HtmlCodeResult):
        if result.html_code:
            files["index.html"] = result.html_code
    elif isinstance(result, MultiFileCodeResult):
        if result.html_code:
            files["index.html"] = result.html_code
        if result.css_code:
            files["style.css"] = result.css_code
        if result.js_code:
            files["script.js"] = result.js_code
    return files
